File: aoc/2025/aoc202507.hs
module AOC202507 where

import Data.Map(alter,empty,insert,member,toList,(!))
import Data.Set(elems,fromList,size)

import AOC

aoc = AOC {
    day="07",
    aocTests=[
        AOCTest {
            testData=unlines [
                ".......S.......",
                "...............",
                ".......^.......",
                "...............",
                "......^.^......",
                "...............",
                ".....^.^.^.....",
                "...............",
                "....^.^...^....",
                "...............",
                "...^.^...^.^...",
                "...............",
                "..^...^.....^..",
                "...............",
                ".^.^.^.^.^...^.",
                "..............."
            ],
            testResult=Just "21",
            testResult2=Just "40"
            }
        ],
    aocCode=Code {
        codeParse=parse2d,
        codeParse2=parse2d,
        codeTest=result,
        codeTest2=result2,
        codeResult=result,
        codeResult2=result2
        }
    }

result manifold = size $ fromList splits
  where
    splits = sendBeams (map fst $ filter ((== 'S') . snd) $ toList manifold)

    sendBeams beams
      | null nextBeams = newSplits
      | otherwise = newSplits ++ sendBeams (elems $ fromList nextBeams)
      where
        (newSplits,nextBeams) = foldr sendBeam ([],[]) beams
    sendBeam (x,y) (splitAcc,beamAcc)
      | not $ member (x,y) manifold = (splitAcc,beamAcc)
      | manifold!(x,y) == 'S' = (splitAcc,(x,y+1):beamAcc)
      | manifold!(x,y) == '.' = (splitAcc,(x,y+1):beamAcc)
      | manifold!(x,y) == '^' = ((x,y):splitAcc,(x-1,y):(x+1,y):beamAcc)

result2 manifold = sum $ sendBeams $ mergeBeams $ map (fmap (const 1))
                                   $ filter ((== 'S') . snd) $ toList manifold
  where
    sendBeams beams
      | null nextBeams = beams
      | otherwise = sendBeams $ mergeBeams nextBeams
      where
        nextBeams = concatMap sendBeam $ toList beams
    sendBeam ((x,y),n)
      | not $ member (x,y+1) manifold = []
      | manifold!(x,y+1) == '.' = [((x,y+1),n)]
      | manifold!(x,y+1) == '^' = [((x-1,y+1),n),((x+1,y+1),n)]
    mergeBeams = foldr merge empty
      where
        merge (xy,n) nextBeams = alter (Just . maybe n (n+)) xy nextBeams

File: aoc/2025/aoc202507.go
package main

func init() {
	Register(&aoc202507{
		AOC: AOC{
			Day: 7,
			Tests: []Test{
				Test{
					Input: `.......S.......
...............
.......^.......
...............
......^.^......
...............
.....^.^.^.....
...............
....^.^...^....
...............
...^.^...^.^...
...............
..^...^.....^..
...............
.^.^.^.^.^...^.
...............
`,
					Part1: "21",
					Part2: "40",
				},
			},
		},
	})
}

type aoc202507 struct {
	AOC
}

func (aoc *aoc202507) Part1(input *Input) string {
	result := 0
	width, height, manifold := input.Grid()
	beams := map[[2]int]bool{}
	for y := range height {
		for x := range width {
			switch manifold[[2]int{x, y}] {
			case 'S':
				beams[[2]int{x, y}] = true
			case '.':
				if beams[[2]int{x, y - 1}] ||
					(manifold[[2]int{x - 1, y}] == '^' && beams[[2]int{x - 1, y - 1}]) ||
					(manifold[[2]int{x + 1, y}] == '^' && beams[[2]int{x + 1, y - 1}]) {
					beams[[2]int{x, y}] = true
				}
			}
			if beams[[2]int{x, y}] && manifold[[2]int{x, y + 1}] == '^' {
				result++
			}
		}
	}
	return IntResult(result)
}

func (aoc *aoc202507) Part2(input *Input) string {
	width, height, manifold := input.Grid()
	beams := map[[2]int]int{}
	for y := range height {
		for x := range width {
			switch manifold[[2]int{x, y}] {
			case 'S':
				beams[[2]int{x, y}] = 1
			case '.':
				timelines := beams[[2]int{x, y - 1}]
				if manifold[[2]int{x - 1, y}] == '^' {
					timelines += beams[[2]int{x - 1, y - 1}]
				}
				if manifold[[2]int{x + 1, y}] == '^' {
					timelines += beams[[2]int{x + 1, y - 1}]
				}
				beams[[2]int{x, y}] = timelines
			}
		}
	}
	result := 0
	for x := range width {
		result += beams[[2]int{x, height - 1}]
	}
	return IntResult(result)
}
